core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -328,6 +366,202 @@ public void close() throws Exception {
         // TODO: Provide Implementation
     }
 
+    private ShareSessionKey shareSessionKey(String groupId, Uuid memberId) {
+        return new ShareSessionKey(groupId, memberId);
+    }
+
+    private static String partitionsToLogString(Collection<TopicIdPartition> partitions) {
+        return FetchSession.partitionsToLogString(partitions, log.isTraceEnabled());
+    }
+
+    /**
+     * Recursive function to process all the fetch requests present inside the fetch queue
+     */
+    private void maybeProcessFetchQueue() {
+        if (!processFetchQueueLock.compareAndSet(false, true)) {
+            // The queue is already being processed hence avoid re-triggering.
+            return;
+        }
+
+        // Initialize the topic partitions for which the fetch should be attempted.
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = new LinkedHashMap<>();
+        ShareFetchPartitionData shareFetchPartitionData = fetchQueue.poll();
+        if (shareFetchPartitionData == null) {
+            // No more requests to process, so release the lock. Though we should not reach here as the lock
+            // is acquired only when there are requests in the queue. But still, it's safe to release the lock.
+            releaseProcessFetchQueueLock();
+            return;
+        }
+
+        try {
+            shareFetchPartitionData.topicIdPartitions.forEach(topicIdPartition -> {
+                SharePartitionKey sharePartitionKey = sharePartitionKey(
+                    shareFetchPartitionData.groupId,
+                    topicIdPartition
+                );
+                SharePartition sharePartition = partitionCacheMap.computeIfAbsent(sharePartitionKey,
+                    k -> new SharePartition(shareFetchPartitionData.groupId, topicIdPartition, maxInFlightMessages, maxDeliveryCount,
+                        recordLockDurationMs, timer, time, persister));
+                int partitionMaxBytes = shareFetchPartitionData.partitionMaxBytes.getOrDefault(topicIdPartition, 0);
+                // Add the share partition to the list of partitions to be fetched only if we can
+                // acquire the fetch lock on it.
+                if (sharePartition.maybeAcquireFetchLock()) {
+                    // If the share partition is already at capacity, we should not attempt to fetch.
+                    if (sharePartition.canAcquireRecords()) {
```

Comment:
sharePartition.releaseFetchLock() doesn't seem to be called for this `if` block.

Reply:
Thanks for looking into this. If the lock is attained on the partition then it ll be released later once fetch proceeds. Either of the 3 places below.

https://github.com/apache/kafka/blob/adee6f0cc11c37edaf7b47c1f77457d66e80ebe4/core/src/main/java/kafka/server/share/SharePartitionManager.java#L435
https://github.com/apache/kafka/blob/adee6f0cc11c37edaf7b47c1f77457d66e80ebe4/core/src/main/java/kafka/server/share/SharePartitionManager.java#L462
https://github.com/apache/kafka/blob/adee6f0cc11c37edaf7b47c1f77457d66e80ebe4/core/src/main/java/kafka/server/share/SharePartitionManager.java#L474